Add nodejs helm chart (#42)
redhat-developer/redhat-helm-charts: alpha/nodejs-chart/README.md

```yaml
build:
  pullSecret: my-pull-secret
```
and apply by passing `--values $VALUES_FILE`.

## Values
This section describes the Values used to configure this chart.

Below is a table of each value used to configure this chart.

| Value | Description | Default | Additional Information |
| ----- | ----------- | ------- | ---------------------- |
| `image.name` | Name of the image you want to build/deploy | Defaults to the Helm release name. | The chart will create/reference an [ImageStream](https://docs.openshift.com/container-platform/4.6/openshift_images/image-streams-manage.html) based on this value. |
| `image.tag` | Tag that you want to build/deploy | `latest` | The chart will create/reference an [ImageStreamTag](https://docs.openshift.com/container-platform/4.6/openshift_images/image-streams-manage.html#images-using-imagestream-tags_image-streams-managing) based on the name provided |
| `build.enabled` | Determines if build-related resources should be created. | `true` | Set this to `false` if you want to deploy a previously built image. Leave this set to `true` if you want to build and deploy a new image. |

> | `build.uri` | Git URI that references your git repo | https://github.com/nodeshift-starters/nodejs-rest-http | This value defaults to a sample application. Be sure to override this if you want to build and deploy your own application. |

Reviewer: @sbose78 Do we want to move this quickstart under https://github.com/redhat-developer-helm-quickstarts to be consistent with Quarkus?

Reviewer 2: Given that https://github.com/nodeshift-starters is a popular location where repos are maintained by Red Hat, we may leave it there for now.

Reviewer 3: +1 on pointing to the example where it is. If `redhat-developer-helm-quickstarts` is going to hold a number of examples at some point we can considered putting a quick start specifically for helm using Node.js later on.

> | `build.ref` | Git ref containing the application you want to build | master | - |

Reviewer: Let's change the default to `main` to be consistent with the quarkus chart

# Node.js Helm Chart
A Helm chart for building and deploying a [Node.js](https://nodejs.org/) application on OpenShift.

## Prerequisites
You will need to create a pull secret if you pull an S2I builder or Docker base image from an external registry. Use the following command as a reference to create your pull secret:
```bash
oc create secret docker-registry my-pull-secret --docker-server=$SERVER_URL --docker-username=$USERNAME --docker-password=$PASSWORD --docker-email=$EMAIL
```

You can use this secret by passing `--set build.pullSecret=my-pull-secret` to `helm install`, or you can configure this in a values file:
```yaml
build:
  pullSecret: my-pull-secret
```
and apply by passing `--values $VALUES_FILE`.

## Values
This section describes the Values used to configure this chart.

### Required Values 
Below shows the required values to install this chart.

| Value | Description | Default | Additional Information |
| ----- | ----------- | ------- | ---------------------- |

> | `build.uri` | Git URI that references your git repo | - | **REQUIRED** |

Reviewer: This isn't yet merged in for the Quarkus chart, but #40 sets a default build.uri for Quarkus so users can just install right off the shelf to build a quickstart application. We're going to want to do this here as well.

### All Values
Below is a table of each value used to configure this chart.

| Value | Description | Default | Additional Information |
| ----- | ----------- | ------- | ---------------------- |
| `image.name` | Name of the image you want to build/deploy | Defaults to the Helm release name. | The chart will create/reference an [ImageStream](https://docs.openshift.com/container-platform/4.6/openshift_images/image-streams-manage.html) based on this value. |
| `image.tag` | Tag that you want to build/deploy | `latest` | The chart will create/reference an [ImageStreamTag](https://docs.openshift.com/container-platform/4.6/openshift_images/image-streams-manage.html#images-using-imagestream-tags_image-streams-managing) based on the name provided |
| `build.enabled` | Determines if build-related resources should be created. | `true` | Set this to `false` if you want to deploy a previously built image. Leave this set to `true` if you want to build and deploy a new image. |
| `build.uri` | Git URI that references your git repo | - | **REQUIRED** |

> | `build.ref` | Git ref containing the application you want to build | master | - |

Reviewer: In PR #40 I'm changing the default to `main` since this is the new default branch for new repos

| `build.contextDir` | The sub-directory where the application source code exists | - | - |

> | `build.mode` | Determines whether the Quarkus mode should be set to `jvm` or `native` | `jvm` | Options: `jvm` or `native` |

Reviewer: Does a `build.mode` value make sense in the context of NodeJS? Or should all builds just default to using our nodejs s2i builder

| `deploy.livenessProbe` | Freeform `livenessProbe` stanza. | See [values.yaml](./values.yaml) | More information: https://docs.openshift.com/container-platform/4.6/applications/application-health.html#application-health-about_application-health |
| `deploy.readinessProbe` | Freeform `readinessProbe` stanza. | See [values.yaml](./values.yaml) | More information: https://docs.openshift.com/container-platform/4.6/applications/application-health.html#application-health-about_application-health |
| `deploy.env` | Freeform `env` stanza | - | More information: https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/ |
| `deploy.envFrom` | Freeform `envFrom` stanza | - | More information: https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#configure-all-key-value-pairs-in-a-configmap-as-container-environment-variables |
| `deploy.applicationProperties.enabled` | Determines if application properties should be externalized in a ConfigMap | `false` | - |
| `deploy.applicationProperties.mountPath` | Location to mount the application.properties file | `/deployments/config/` | - |

> | `deploy.applicationProperties.properties` | The application.properties file contents | - | - |

Reviewer: These `applicationProperties` values are Quarkus-specific. However, if there's a good way to externalize nodejs properties, these could possibly be repurposed with a different name

# Node.js Helm Chart
A Helm chart for building and deploying a [Node.js](https://nodejs.org/) application on OpenShift.

## Prerequisites
Below are prerequisites that may apply to your use case.

### Pull Secret

> You will need to create a pull secret if you pull an S2I builder or Docker base image from an external registry. Use the following command as a reference to create your pull secret:

Reviewer 4: Not required when making use of `ImageStreams` in the `openshift` namespace

```bash
oc create secret docker-registry my-pull-secret --docker-server=$SERVER_URL --docker-username=$USERNAME --docker-password=$PASSWORD --docker-email=$EMAIL
```

You can use this secret by passing `--set build.pullSecret=my-pull-secret` to `helm install`, or you can configure this in a values file:
```yaml
build:
  pullSecret: my-pull-secret
```
and apply by passing `--values $VALUES_FILE`.

### Push Secret
You will need to create a push secret if you want to push your image to an external registry. Use the following command as a reference to create your push secret:
```bash
oc create secret docker-registry my-push-secret --docker-server=$SERVER_URL --docker-username=$USERNAME --docker-password=$PASSWORD --docker-email=$EMAIL
```

You can use this secret by passing `--set build.output.pushSecret=my-push-secret` and `--set build.output.kind=DockerImage` to `helm install`, or you can configure these in a values file:
```yaml
build:
  output:
    kind: DockerImage
    pushSecret: my-push-secret
```

## Values
This section describes the Values used to configure this chart.

> Below is a table of each value used to configure this chart.

Reviewer 3: ```suggestion
Below is a table the values used to configure this chart.
```

| `build.enabled` | Determines if build-related resources should be created. | `true` | Set this to `false` if you want to deploy a previously built image. Leave this set to `true` if you want to build and deploy a new image. |
| `build.uri` | Git URI that references your git repo | https://github.com/nodeshift-starters/nodejs-rest-http | This value defaults to a sample application. Be sure to override this if you want to build and deploy your own application. |
| `build.ref` | Git ref containing the application you want to build | main | - |
| `build.contextDir` | The sub-directory where the application source code exists | - | - |
| `build.output.kind` | Determines if the image will be pushed to an ImageStreamTag or a DockerImage (external registry) | ImageStreamTag | More information: More information: https://docs.openshift.com/container-platform/4.6/builds/managing-build-output.html |
| `build.output.pushSecret` | Push secret name | - | Used only if build.output.kind == 'DockerImage' |

> | `build.pullSecret` | Image pull secret | - | More information: https://docs.openshift.com/container-platform/4.6/openshift_images/managing_images/using-image-pull-secrets.html |

Reviewer: @AshCripps Can you add `build.output.kind` and `build.output.pushSecret` to the values.schema.json also? Apologies, missed this earlier. 

Here's an example from the Quarkus chart - https://github.com/redhat-developer/redhat-helm-charts/blob/master/alpha/quarkus-chart/values.schema.json#L80-L94

| `build.env` | Freeform `env` stanza | - | More information: https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/ |
| `build.resources` | Freeform `resources` stanza | - | More information: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/ |
| `deploy.replicas` | Number of pod replicas to deploy | `1` | - |
| `deploy.resources` | Freeform `resources` stanza | - | More information: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/ |
| `deploy.serviceType` | Type of service to create | `ClusterIP` | More information: https://kubernetes.io/docs/concepts/services-networking/service/#publishing-services-service-types |

> | `deploy.ports` | Freeform service `ports` stanza. | See [values.yaml](./values.yaml) | More information: https://kubernetes.io/docs/concepts/services-networking/service/#defining-a-service |

Reviewer 3: I'm on the fence whether a reference to the values.yaml versus just including the default info in-line

Reviewer: I felt that less-technical users would find it helpful in the readme, but get what you're saying if you're trying to avoid duplicating the reference in values.yaml and readme.

Reviewer 3: I was thinking it might make sense to pull in versus the reference to the values.yaml, but don't feel to strongly one way or the other.
